contracts/src/CrossChainApplications/NativeTokenBridge/NativeTokenSource.sol

```diff
@@ -0,0 +1,201 @@
+        } else {
+            revert("NativeTokenSource: invalid action");
+        }
+    }
+
+    /**
+     * @dev See {INativeTokenSource-transferToDestination}.
+     */
+    function transferToDestination(
+        address recipient,
+        TeleporterFeeInfo calldata feeInfo,
+        address[] calldata allowedRelayerAddresses
+    ) external payable nonReentrant {
+        // The recipient cannot be the zero address.
+        require(
+            recipient != address(0),
+            "NativeTokenSource: zero ercipient address"
+        );
+
+        // Lock tokens in this bridge instance. Supports "fee/burn on transfer" ERC20 token
+        // implementations by only bridging the actual balance increase reflected by the call
+        // to transferFrom.
+        uint256 adjustedFeeAmount = 0;
+        if (feeInfo.amount > 0) {
+            adjustedFeeAmount = SafeERC20TransferFrom.safeTransferFrom(
+                IERC20(feeInfo.feeTokenAddress),
+                feeInfo.amount
+            );
+            SafeERC20.safeIncreaseAllowance(
+                IERC20(feeInfo.feeTokenAddress),
+                address(teleporterMessenger),
+                adjustedFeeAmount
+            );
+        }
+
+        uint256 messageID = teleporterMessenger.sendCrossChainMessage(
+            TeleporterMessageInput({
+                destinationChainID: destinationBlockchainID,
+                destinationAddress: nativeTokenDestinationAddress,
+                feeInfo: feeInfo,
+                requiredGasLimit: MINT_NATIVE_TOKENS_REQUIRED_GAS,
+                allowedRelayerAddresses: allowedRelayerAddresses,
+                message: abi.encode(recipient, msg.value)
+            })
+        );
+
+        emit TransferToDestination({
+            sender: msg.sender,
+            recipient: recipient,
+            amount: msg.value,
+            teleporterMessageID: messageID
+        });
+    }
+
+    /**
+     * @dev Unlocks tokens to recipient.
+     */
+    function _unlockTokens(address recipient, uint256 amount) private {
+        require(
+            recipient != address(0),
+            "ERC20TokenSource: zero recipient address"
+        );
+
+        // Transfer to recipient
+        payable(recipient).transfer(amount);
+
+        emit UnlockTokens(recipient, amount);
+    }
+
+    /**
+     * @dev Sends tokens to BLACKHOLE_ADDRESS.
+     */
+    function _burnTokens(uint256 amount) private {
+        payable(BURNED_TX_FEES_ADDRESS).transfer(amount);
+        emit BurnTokens(amount);
+    }
+
+    /**
+     * @dev Update destinationChainBurnedBalance sent from destination chain
+     */
+    function _updateDestinationChainBurnedBalance(
+        uint256 newBurnBalance
+    ) private {
+        if (newBurnBalance > destinationChainBurnedBalance) {
+            uint256 difference = newBurnBalance - destinationChainBurnedBalance;
+            _burnTokens(difference);
+            destinationChainBurnedBalance = newBurnBalance;
+        }
```

Comment:
Does newBurnBalance have to be >= destinationChainBurnedBalance? Is this something we should check for?

Reply:
It doesn't HAVE to be. The burned balance on the destination chain is monotonically increasing, but two calls to the report function could be received out of order. We only want to burn tokens when the number we receive is higher than the highest we've seen.